pkg/util/pki/generate.go

```diff
@@ -1,18 +1,70 @@
 package pki
 
 import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
 	"crypto/rand"
 	"crypto/rsa"
 	"crypto/x509"
 	"encoding/pem"
+	"fmt"
 )
 
+var ellipticCurves = map[int]elliptic.Curve{
+	224: elliptic.P224(),
+	256: elliptic.P256(),
+	384: elliptic.P384(),
+	521: elliptic.P521(),
+}
+
+func GeneratePrivateKey(algorithm string, keySize int) (crypto.PrivateKey, error) {
+	switch algorithm {
+	case "rsa":
+		return GenerateRSAPrivateKey(keySize)
+	case "ecdsa":
+		return GenerateECPrivateKey(keySize)
+	default:
+		return nil, fmt.Errorf("unsupported private key algorithm specified: %s", algorithm)
+	}
+}
+
 func GenerateRSAPrivateKey(keySize int) (*rsa.PrivateKey, error) {
 	return rsa.GenerateKey(rand.Reader, keySize)
 }
 
+func GenerateECPrivateKey(keySize int) (*ecdsa.PrivateKey, error) {
+	ecCurve, ok := ellipticCurves[keySize]
+	if !ok {
+		return nil, fmt.Errorf("unsupported ecdsa key size specified: %d", keySize)
+	}
+
+	return ecdsa.GenerateKey(ecCurve, rand.Reader)
+}
+
```

Comment:
The addition of EC keys is great, but can this be extracted out into a separate PR? It will take additional work to support this for all Issuer types.

Reply:
Cool. I've extracted this work to PR https://github.com/jetstack/cert-manager/pull/722